### ISSUE TITLE
Possibly missing dsl construct for single-valued child attribute

### ISSUE THREAD

**Issue by strindberg** [jOOQ/jOOQ#12479]
This is, in a way, a continuation of #12347. I have now realized that what caused some confusion in the reported example was the fact that our code was a mixture of mappings using reflection/jackson and the new functional features in the most recent jOOQ release.

It could very well be that I am missing something, but it seems to me that the new functional approach could be enhanced when mapping single-value child objects (i.e. not collections).

I have prepared an example MCVE here: https://github.com/strindberg/jOOQ-mcve, branch `jooq_issue_mapping`. The test code contains two methods for reading an object and nested sub-ojbects from the database. The method `fun getByIdFunctional(id: Long): Parent?` contains an attempt to only use the new functional mapping possibilities.

The problem that concerns me is the reading of the `owner` attribute. In previous jOOQ releases we used the type of jsonObject wrapping that you can see in `fun getByIdWithReflection(id: Long)`. Now, when we try to avoid json mapping and reflection, I see no simpler way of mapping the `owner` object than to use `multiselect` and then explicitly taking the first object from the resulting list. Perhaps this is the way it must be, but it seems kind of clunky: the owner sub-object is no a list, but a single value in our data model.

I am specifically talking about this code:

```kotlin
    fun getByIdFunctional(id: Long): Parent? =
        ctx().select(
            PARENT.ID,
            PARENT.NAME,
            multiset(
                select(OWNER.ID, OWNER.FIRST_NAME, OWNER.LAST_NAME)
                    .from(OWNER).where(OWNER.PARENT_ID.eq(PARENT.ID))
            ).convertFrom { it[0].run { Owner(value1()!!, value2()!!, value3()!!) } },
...
```

Is there already a way to express this so that it more directly maps the sub object to only a single value?

**Comment by lukaseder** [comment]
You're looking for simple nested records, not multisets, see:
https://www.jooq.org/doc/latest/manual/sql-building/column-expressions/nested-records/

Hope this helps

**Comment by strindberg** [comment]
I am sorry, but I fail to see how nested records solves my problem in this case: the attributes in the `Owner` object comes from the `OWNER` table , not the `PARENT`table, and I of course need to specify which row in the `OWNER` table should be mapped into the `Owner` object. This is present in the `select`statement within the `multiset` in my example, but I don't see how the same would be expressed within a `row` statement.

**Comment by lukaseder** [comment]
Here's an option:

```kotlin
    fun getByIdFunctional(id: Long): Parent? =
        ctx().select(
            PARENT.ID,
            PARENT.NAME,
            row(OWNER.ID, OWNER.FIRST_NAME, OWNER.LAST_NAME).mapping(::Owner),
            multiset(
                select(CHILD.ID, CHILD.NAME)
                    .from(CHILD).where(CHILD.PARENT_ID.eq(PARENT.ID))
            ).convertFrom { it.map { s -> Child(s.value1()!!, s.value2()!!) } },
        )
            .from(PARENT)
            .join(OWNER).on(OWNER.PARENT_ID.eq(PARENT.ID)) // Joins are still a thing, even with MULTISET ;-)
            .where(PARENT.ID.eq(id))
            .fetchOne { r -> Parent(r.value1()!!, r.value2()!!, r.value3(), r.value4()!!) }
```

Though, I wonder if the `PARENT` -> `OWNER` relationship is in the right direction? Is it 1:1?

**Comment by strindberg** [comment]
OK, I see. Thank you for taking the time to respond. I should have called Owner something else: I was indeed interested in expressing a 1:1 relationship, but the name Owner suggests something else.

But wasn't the point of using jsonObject in a case such as this precisely to avoid the join? Doesn't this mean that there is indeed a gap in functionality here? Either we use jsonObject to avoid the join but lose the type safety, or we use row(...) and are forced to use the join but keep the type safety? We can't have both?

By the way, apart from the fact that using multiset to read a single value looks a bit weird, is there any real drawbacks to that approach?

**Comment by lukaseder** [comment]
> But wasn't the point of using jsonObject in a case such as this precisely to avoid the join?

Why? There's no inherently "good" or "bad" relational operator. There are many tools that can be used to achieve a thing. Which is "better"?

- `t.a IN (SELECT u.b FROM u)` 
- `EXISTS (SELECT * FROM u WHERE t.a = u.b)`

The two above are logically equivalent. Most optimisers will also produce the same execution plans. Depending on the constraints you have on those tables, there might be even an equivalent `INNER JOIN` that some optimisers can transform to the above semi joins.

With nesting operators, there may be a bigger difference in what optimisers currently produce (and there might not be exact logical equivalence), though this might or might not matter for you.

> Doesn't this mean that there is indeed a gap in functionality here?

A subquery produces a multiset of data, but you didn't want a multiset of data, you wanted a single row. You only cared about the first item of your multiset, and there's no way to express this with SQL types. Consider the following scalar subquery:

```sql
SELECT (SELECT 1)
```

The subquery is required to produce 1 column x 1 row. The 1 column constraint can be type checked, the 1 row constraint cannot. It is validated at runtime. In jOOQ, this is done using `field(Select)` for scalar subqueries. You can of course create a `field(Select)` for non-scalar subqueries too, if you're projecting rows. I wanted to omit that syntax because I didn't have an example from the manual ready. But since jOOQ 3.15, `Row3<T1, T2, T3>` extends `SelectField<Record3<T1, T2, T3>>`, so you can avoid the join again, if you feel that's important for readability reasons, etc., or if you notice a performance difference.

> Either we use jsonObject to avoid the join but lose the type safety

The point of using `jsonObject` starting from jOOQ 3.14 was twofold:

- The whole JSON/XML infrastructure needed to be implemented *before* the new `MULTISET` support, and it needed to be robust, first.
- It happened to be a good approximation of using `MULTISET` even before `MULTISET` was available, because nesting documents as `XML`, `JSON`, or `MULTISET` is all the same thing, just a different way to serialise them over the wire.

So, at the time of jOOQ 3.14, I recommended using the JSON approach first, as it already worked very well (with Jackson) and offered a lot of value. Starting from jOOQ 3.15, it's probably not the best approach (though it still works well), because it doesn't offer type safety, and leaves the burden of mapping to the user, in cases where Jackson/Gson don't work well out of the box.

I've updated almost all Stack Overflow answers to recommend `MULTISET` over `JSON_OBJECT` now.

> or we use row(...) and are forced to use the join but keep the type safety? We can't have both?

You can do anything you want, all of these are just operators, and they compose very well (unless there's a bug in jOOQ, or a limitation in the RDBMS, of course). You can have both.

> By the way, apart from the fact that using multiset to read a single value looks a bit weird, is there any real drawbacks to that approach?

Yeah, it does look a bit weird. At least, it's confusing to the extent where it doesn't clearly communicate the intent of the query. That's a flaw of SQL that jOOQ inherits. There isn't really a way of distinguishing 0..1 collections from 0..N collections, it's all the same (thus the example with the scalar subquery `SELECT (SELECT 1)`). There's also not really a way to express a 1:1 relationship, all relationships are 0..N, so a join can't guarantee there won't be a cartesian product, unless you're joining from child to parent. If that were the case in your actual schema, you could be using the implicit join syntax, which would make this a bit more readable, depending on your taste:

```kotlin
    fun getByIdFunctional(id: Long): Parent? =
        ctx().select(
            PARENT.ID,
            PARENT.NAME,
            // Implicit join here from PARENT to OWNER via PARENT.owner()
            row(PARENT.owner().ID, PARENT.owner().FIRST_NAME, PARENT.owner().LAST_NAME).mapping(::Owner),
            multiset(
                select(CHILD.ID, CHILD.NAME)
                    .from(CHILD).where(CHILD.PARENT_ID.eq(PARENT.ID))
            ).convertFrom { it.map { s -> Child(s.value1()!!, s.value2()!!) } },
        )
            .from(PARENT)
            .where(PARENT.ID.eq(id))
            .fetchOne { r -> Parent(r.value1()!!, r.value2()!!, r.value3(), r.value4()!!) }
```

So, in other words, when using `MULTISET` for single row subqueries, the main drawback is a lack of being able to express the 0..1 or 1..1 arity. But you can add a comment, or maybe it's clear from the names of the actual tables. Or, you use any of the alternative syntaxes...
